stdlib/public/core/EmbeddedRuntime.swift

```diff
@@ -0,0 +1,121 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftShims
+
+public struct ClassMetadata {
+  var superclassMetadata: UnsafeMutablePointer<ClassMetadata>?
+
+  // There is no way to express the actual calling convention on the heap desroy
+  // function (swiftcc with 'self') currently, so let's use UnsafeRawPointer
+  // and a helper function in C (_swift_runtime_invoke_heap_object_destroy).
+  var destroy: UnsafeRawPointer
+}
+
+public struct HeapObject {
+  var metadata: UnsafeMutablePointer<ClassMetadata>
+  var refcount: Int
```

Comment:
I know this is just an initial "experiment". For the final version we need to think about supporting unowned and/or weak references.

Reply:
Yes. Let me write that down as a TODO.